### PR TITLE
Fix #7: Implement DeferredDispatcher

### LINKED ISSUE #7: Consider introducing deferred event provider
Sometimes we use events in transactions, so they should not be dispatched before the transaction is successfully committed. As an example, imagine a user registration flow:

1. Validate data
2. Begin transaction
3. Write user to DBMS
4. Dispatch `UserWasCreated` event
5. Create a profile, catch an exception
6. Rollback transaction

The dependent services may have already sent the welcome email as a result of `UserWasCreated` event handling, but the transaction was reverted later. 

That's why we use an Emitter – a thing that collects events but does not send them to a Dispatcher until `flush()` is called.

The implementation can be pretty dumb:

```php
interface EventDispatcherInterface
{
    public function flush(): void;
}

final class Emitter implements EventDispatcherInterface
{
    private $dispatcher;
    private $events = [];

    public function __construct(EventDispatcherInterface $dispatcher)
    {
        $this->dispatcher = $dispatcher;
    }

    public function dispatch(object $event)
    {
         return $this->events[] = $event;
    }

    public function flush(): void
    {
        foreach ($this->events as $event) {
            $dispatcher->dispatch($event);
        }
    }
}
```

```php
$provider = new Yiisoft\EventDispatcher\Provider\Provider();
$provider->attach(function (UserWasCreated $event) {
    sendGreetingEmail();
});

$dispatcher = new Yiisoft\EventDispatcher\Dispatcher($provider);
$emitter = new Yiisoft\EventDispatcher\Emitter($dispatcher);

$emitter->dispatch(new UserWasCreated($uuid));
// ... flow contiues
$emitter->flush();
```

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #7

### CODE REVIEW

**Comment by yiiliveext** [comment]
How are you going to use it with non-deferred events?

File: src/Dispatcher/DeferredDispatcher.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    public function __construct(EventDispatcherInterface $dispatcher)
+    {
+        $this->dispatcher = $dispatcher;
+    }
+
+
+    public function dispatch(object $event)
+    {
+        $this->events[] = $event;
```
Comment (vjik) [inline src/Dispatcher/DeferredDispatcher.php:20]: `dispath` should return event by [interface](https://github.com/php-fig/event-dispatcher/blob/3ef040d7c05652f55bd05115baf059a445cfc79c/src/EventDispatcherInterface.php#L17).

**Comment by samdark** [comment]
@yiiliveext do you mean events that mutate state? We can't defer these by defintion.

**Comment by yiiliveext** [comment]
> @yiiliveext do you mean events that mutate state? We can't defer these by definition.

Do you want to create `DeferredDispoatcher` every time you need?

**Comment by samdark** [comment]
Good idea. Done.

**Comment by vjik** [comment]
Maybe on flush return dispatched events?  May need them for further processing...

**Comment by samdark** [comment]
Good idea. Done.

File: src/Dispatcher/DeferredDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+
```
Comment (rustamwin) [inline src/Dispatcher/DeferredDispatcher.php:2]:
```suggestion
declare(strict_types=1);

```

**Review by rustamwin — APPROVED** [review]

**Comment by samdark** [comment]
After some thinking and @roxblnfk comments about possible state issues I've decided not to have it.

**Pull request closed without merging**